batchproducer/batchproducer.go

```diff
@@ -0,0 +1,314 @@
+// or greater than bufferSize then the Add method will block.
+// TODO: this is too many args. Maybe instead have some defaults and add some setter methods for the less critical
+// params
+func NewBatchProducer(
+	client BatchingKinesisClient,
+	streamName string,
+	bufferSize int,
+	flushInterval time.Duration,
+	batchSize int,
+	logger *log.Logger,
+) (BatchProducer, error) {
+	if batchSize < 1 || batchSize > 500 {
+		return nil, errors.New("batchSize must be between 1 and 500 inclusive")
+	}
+
+	if bufferSize < batchSize && flushInterval <= 0 {
+		return nil, errors.New("If bufferSize < batchSize && flushInterval <= 0 then the buffer will eventually fill up and Add will block forever.")
+	}
+
+	if flushInterval > 0 && flushInterval < 50*time.Millisecond {
+		return nil, errors.New("Are you crazy?")
+	}
+
+	batchProducer := batchProducer{
+		client:               client,
+		streamName:           streamName,
+		flushInterval:        flushInterval,
+		batchSize:            batchSize,
+		maxAttemptsPerRecord: 10,
+		logger:               logger,
+		statInterval:         time.Second,
+		currentStat:          new(BatchStat),
+		records:              make(chan batchRecord, bufferSize),
+		stop:                 make(chan interface{}),
+	}
+
+	return &batchProducer, nil
+}
+
+type batchProducer struct {
+	client               BatchingKinesisClient
+	streamName           string
+	flushInterval        time.Duration
+	batchSize            int
+	maxAttemptsPerRecord int
+	logger               *log.Logger
+	running              bool
+	consecutiveErrors    int
+	currentDelay         time.Duration
+	statInterval         time.Duration
+	statReceiver         BatchStatReceiver
+	currentStat          *BatchStat
+	records              chan batchRecord
+	stop                 chan interface{}
+}
+
+type batchRecord struct {
+	data         []byte
+	partitionKey string
+	sendAttempts int
+}
+
+// from/for interface BatchProducer
+func (b *batchProducer) Add(data []byte, partitionKey string) error {
+	if !b.running {
+		return errors.New("Cannot call Add when BatchProducer is not running (to prevent the buffer filling up and Add blocking indefinitely).")
+	}
+	b.records <- batchRecord{data: data, partitionKey: partitionKey}
+	return nil
+}
+
+// from/for interface BatchProducer
+func (b *batchProducer) Start() error {
+	if b.running {
+		return nil
+	}
+
+	go b.run()
+
+	for !b.running {
+		// Give the goroutine a chance to start before returning.
+		time.Sleep(1 * time.Microsecond)
+	}
+
+	return nil
+}
+
+func (b *batchProducer) run() {
+	flushTick := time.Tick(b.flushInterval)
+
+	var statTick <-chan time.Time
+	if b.statReceiver != nil {
+		statTick = time.Tick(b.statInterval)
+	}
+
+	b.running = true
+	defer func() { b.running = false }()
+
+	for {
+		select {
+		case <-flushTick:
+			b.sendBatch()
+		case <-statTick:
+			b.sendStats()
+		case <-b.stop:
+			break
+		default:
+			if len(b.records) >= b.batchSize {
```

Comment:
What is reading `b.records`?

We could have `Add` trigger a `chan` that can be read off of:

```go
select {
case <-flushTick:
	b.sendBatch()
case <-b.full:
	b.sendBatch()
case <-statTick:
	b.sendStats()
case <-b.stop:
	break
  // snip
}
```

This way, you don't have to poll (which is essentially what is happening with the `time.Sleep`